js/cryptopia.js

```diff
@@ -728,22 +716,59 @@ module.exports = class cryptopia extends Exchange {
     }
 
-    async request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let response = await this.fetch2 (path, api, method, params, headers, body);
-        if (api === 'web')
-            return response;
-        if (response) {
-            if ('Success' in response)
-                if (response['Success']) {
-                    return response;
-                } else if ('Error' in response) {
-                    let error = this.safeString (response, 'error');
-                    if (typeof error !== 'undefined') {
-                        if (error.indexOf ('Insufficient Funds') >= 0)
-                            throw new InsufficientFunds (this.id + ' ' + this.json (response));
+    request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        return this.fetch2 (path, api, method, params, headers, body);
+    }
+
+    nonce () {
+        return this.milliseconds ();
+    }
+
+    handleErrors (code, reason, url, method, headers, body) {
+        if (typeof body !== 'string')
+            return; // fallback to default error handler
+        if (body.length < 2)
+            return; // fallback to default error handler
+        const fixedJSONString = this.sanitizeBrokenJSONString (body);
+        if (fixedJSONString[0] === '{') {
+            let response = JSON.parse (fixedJSONString);
+            if ('Success' in response) {
+                const success = this.safeString (response, 'Success');
+                if (success === 'false') {
+                    let error = this.safeString (response, 'Error');
+                    let feedback = this.id;
+                    if (typeof error === 'string') {
+                        feedback = feedback + ' ' + error;
+                        if (error.indexOf ('does not exist') >= 0) {
+                            throw new OrderNotFound (feedback);
+                        }
+                        if (error.indexOf ('Insufficient Funds') >= 0) {
+                            throw new InsufficientFunds (feedback);
+                        }
+                        if (error.indexOf ('Nonce has already been used') >= 0) {
+                            throw new InvalidNonce (feedback);
+                        }
+                    } else {
+                        feedback = feedback + ' ' + fixedJSONString;
                     }
+                    throw new ExchangeError (feedback);
                 }
+            }
         }
-        throw new ExchangeError (this.id + ' ' + this.json (response));
+    }
+
+    sanitizeBrokenJSONString (jsonString) { // sometimes cryptopia will return a unicode symbol before actual JSON string.
+        const braceCode = 123; // '{'
+        let i = 0;
+        for (i = 0; i < jsonString.length; i++) {
+            if (jsonString.charCodeAt (i) === braceCode) {
```

Comment:
I'm looking for a way to port this line.

Reply:
just find the opening curly brace and cut from there.

Reply:
Or what do you mean by "port" exactly? I can rewrite this to use `indexOf()` or something. Just tell me.

Reply:
Here by "port", I mean, a way to transpile this logical construct to py/php. You see, our transpiler  is regex-based and it is quite dumb and it can only transpile certain statements, so we don't have a regex for charCodeAt because other languages treat strings somewhat differently from JS. We still have to use it, because we don't have a better tool for transpiling yet, and it does the job for us so far... So, if you can rewrite that to indexOf – that would be great! This particular PR is now deprecated, will close it, so, if you will, please do the edits in #3051. Thx again for your really helpful participation!